vital/types/rotation.cxx

```diff
@@ -151,12 +149,11 @@ void
 rotation_< T >
 ::get_yaw_pitch_roll( T& yaw, T& pitch, T& roll ) const
 {
-  Eigen::Matrix< T, 3, 3 > rotM( this->matrix() );
-  T cos_p = T( std::sqrt( double( rotM( 1, 2 ) * rotM( 1, 2 ) ) + rotM( 2, 2 ) * rotM( 2, 2 ) ) );
-
-  yaw   = T( std::atan2( double( rotM( 0, 0 ) ), double( rotM( 0, 1 ) ) ) );
-  pitch = T( std::atan2( double( rotM( 0, 2 ) ), double(cos_p) ) );
-  roll  = T( std::atan2( double( -rotM( 1, 2 ) ), double( -rotM( 2, 2 ) ) ) );
+  roll = std::atan2( 2.0 * ( q_.w() * q_.x() + q_.y() * q_.z() ),
+                     1.0 - 2.0 * ( q_.x() * q_.x() + q_.y() * q_.y() ) );
+  pitch = std::asin( 2.0 * ( q_.w() * q_.y() - q_.x() * q_.z() ) );
+  yaw = std::atan2( 2.0 * ( q_.w() * q_.z() + q_.x() * q_.y() ),
+                    1.0 - 2.0 * ( q_.y() * q_.y() + q_.z() * q_.z() ) );
```

Comment:
Pretty sure there are conversion warnings hiding in here?

```suggestion
  constexpr auto t1 = static_cast< T >( 1.0 );
  constexpr auto t2 = static_cast< T >( 2.0 );
  roll = std::atan2( t2 * ( q_.w() * q_.x() + q_.y() * q_.z() ),
                     t1 - t2 * ( q_.x() * q_.x() + q_.y() * q_.y() ) );
  pitch = std::asin( t2 * ( q_.w() * q_.y() - q_.x() * q_.z() ) );
  yaw = std::atan2( t2 * ( q_.w() * q_.z() + q_.x() * q_.y() ),
                    t1 - t2 * ( q_.y() * q_.y() + q_.z() * q_.z() ) );
```
(Other approaches are possible, this is just the first thing that came to mind. Alternatively, `constexpr T t1 = 1;` is another option along the lines of the above.)

Reply:
Can you explain why `static_cast` is preferred in this case?